lib/IRGen/GenStruct.cpp

```diff
@@ -61,6 +63,47 @@ static StructTypeInfoKind getStructTypeInfoKind(const TypeInfo &type) {
   return (StructTypeInfoKind) type.getSubclassKind();
 }
 
+clang::CXXDestructorDecl *swift::irgen::getDestructor(SILType type) {
+  DestructorDecl *destructor = nullptr;
+  if (auto structDecl = type.getStructOrBoundGenericStruct()) {
+    for (auto member : structDecl->getMembers()) {
+      if (auto dest = dyn_cast<DestructorDecl>(member)) {
+        destructor = dest;
+        break;
+      }
+    }
+  } else if (auto *classDecl = type.getClassOrBoundGenericClass()) {
+    destructor = classDecl->getDestructor();
+  }
+
+  if (!destructor)
+    return nullptr;
+
+  return const_cast<clang::CXXDestructorDecl *>(
+      dyn_cast<clang::CXXDestructorDecl>(destructor->getClangDecl()));
+}
+
+llvm::Constant *
+swift::irgen::getDestructorCallee(IRGenModule &IGM,
+                                  clang::CXXDestructorDecl *destructor) {
+  auto globalDecl = clang::GlobalDecl(destructor, clang::Dtor_Complete);
+  auto clangFnAddr = IGM.getAddrOfClangGlobalDecl(globalDecl, NotForDefinition);
+  return clangFnAddr->stripPointerCasts();
+}
+
+bool swift::irgen::callDestructor(IRGenFunction &IGF, Address address,
+                                  SILType type) {
+  auto *destructor = getDestructor(type);
+  if (!destructor)
+    return false;
+  auto *callee = getDestructorCallee(IGF.IGM, destructor);
+  auto selfTy =
+      cast<llvm::Function>(callee)->getFunctionType()->getParamType(0);
+  auto self = IGF.coerceValue(address.getAddress(), selfTy, IGF.IGM.DataLayout);
+  IGF.Builder.CreateCall(callee, {self});
```

Comment:
C++ destructors can take implicit parameters.

I think the right thing to do here is to use Clang's `CGCXXABI::EmitDestructorCall()` to emit the call. This isn't available through a public Clang API, so you'll need to surface this first -- see here for a similar change related to constructors:

https://reviews.llvm.org/D79942

Reply:
OK. I'll work on a patch for that tomorrow. Thanks for the link.

Reply:
@martinboehme is there a way to get a `CodeGenFunction` associated with the current `llvm::Function`?

Reply:
I don't know, sorry.

Reply:
@martinboehme [Here's the patch](https://reviews.llvm.org/D82392). I tagged you as a reviewer on phab.

Reply:
The patch just landed as e7c5da57a5f3fdb6649ff92cb8ba0ce3c4978668 so, this PR is no longer blocked by that.